vcprompt.1
.TH VCPROMPT 1 "February 2013" "vcprompt" "User Commands"

.SH NAME
vcprompt \- version control information in your prompt

.SH SYNOPSIS
.B vcprompt
[-h] [-d] [-t timeout_ms] [-f format]

.SH DESCRIPTION

Examine the current working directory, determine what version control
system manages it, and print a short summary to stdout. You must
configure your shell to run
.B vcprompt
for every prompt, and your prompt will include that short summary.

By default,
.B vcprompt
just shows the name of the version control system and the current
branch, e.g.
.nf
.in +4m
[git:master]
[hg:default]
[svn:trunk]
[cvs:trunk]
[fossil:trunk]
.in -4m
.fi
This is the default because it's useful, implemented, and fast for
all supported version control systems.

You can get different information by specifying a format string with
the -f option: see \fBFORMAT STRINGS\fR below.

If the current directory is not under version control,
.B vcprompt
prints nothing and exits.

.SH OPTIONS
.IP -d
Print debug messages to stdout, and always end with a newline. Useful
for understanding why
.B vcprompt
does not print what you thought it should print. Do not use this in
your shell prompt!
.IP "-f format"
Specify a custom format string (default: "[%n:%b] "). See \fBFORMAT
STRINGS\fR below.
.IP "-t timeout"
Terminate after
.I timeout
milliseconds of attempting to analyze the working dir. Useful if you
have slow operations (e.g. %m or %u) in your format string that are
tolerable in small working dirs, but not in large ones. The timeout
applies to the whole run of
.B vcprompt.
If it fires, the entire operation fails: no partial information will
be printed.
.IP -F
List features built-in to this
.B vcprompt
binary and terminate (mainly for testing).

.SH FORMAT STRINGS

.B vcprompt
uses printf-style format strings, where % followed by a single letter
expands to some piece of information about the current working dir.
The supported format specifiers are:
.TP
.B %n
A short all-lowercase name for the version control system managing the
working dir: e.g. "git", "hg", "svn", "cvs", "fossil".
.TP
.B %b
The name of the current branch.
.TP
.B %r
The current revision number, changeset ID, or commit ID.
.TP
.B %p
The name of the currently applied patch, if any (Mercurial + MQ only,
but it looks like this could easily be supported for git + guilt).
.TP
.B %u
A single "?" if there are any unknown (untracked) files in the working
dir. Slow.
.TP
.B %m
A single "+" if there are any uncommitted changes (modified, added, or
removed files) in the working dir. Slow.
.TP
.B %%
A single "%" character.
.PP

Not all version control systems support all format specifiers. For
example, CVS has no notion of a single revision identifier, so
.B vcprompt
ignores %r in a CVS working dir. See the section on each version
control system below for more details.

Some format specifiers are slow with most or all supported version
control systems. "Slow" generally means that it requires spawning an
external command, like "git status" or "fossil info".
.B vcprompt
tries hard to avoid spawning external commands, but sometimes it can't
be helped.

.SH GIT SUPPORT

.B vcprompt
considers the current directory a git working dir if directory
.I .git
exists.

.B %b
(branch) is supported by reading
.I .git/HEAD
directly. If that file does not refer to a branch (e.g. you have
checked out a specified commit), then
.B vcprompt
reports the branch as "unknown".

.B %r
(revision) expands to the first 12 characters of the commit ID of
HEAD.

.B %p
is not yet implemented.

.B %u
is supported by running "git ls-files --others --exclude-standard", so
it can be slow in a large working dir.

.B %m
is supported by running "git diff --no-ext-diff --quiet --exit-code",
so it can be slow in a large working dir.

.SH MERCURIAL (HG) SUPPORT

.B vcprompt
considers the current directory a Mercurial (hg) working dir if
directory
.I .hg
exists.

.B %b
(branch) is supported by looking for the current Mercurial bookmark
(\fI.hg/bookmarks.current\fP), or the current named branch
(\fI.hg/branch\fP) if no bookmark is current.

.B %r
(revision) expands to the revision number of the parent of the working
dir, or a comma-separated pair of revision numbers if a merge is
active (the working dir has two parents). If
.B vcprompt
fails to parse some of Mercurial's internal data, it might print a
short changeset ID instead of a revision number. If that happens,
please report a bug in \fBvcprompt\fP!

.B %p
is implemented by reading MQ internals.

.B %u
and
.B %m
are implemented by running "hg status", so they can be slow in a
large working dir. Mercurial has to work harder to find unknown files
than it does to find uncommitted changes, so using
.B %u
can be considerably more expensive than just
.B %m.

.SH SUBVERSION (SVN) SUPPORT

.B vcprompt
considers the current directory a Subversion (svn) working dir if
directory
.I .svn
exists.
.B vcprompt
currently supports Subversion up to version 1.8.

.B %b
is supported by parsing the repository URL corresponding to the root
of your working copy: if we're in a checkout of
.I .../trunk
then branch is "trunk"; if we're in a checkout of
.I .../branches/FOO
then branch is "FOO"; otherwise the branch is unknown.

.B %r
reports the same as "Last Changed Rev" from "svn info" in the root of
your working copy (not necessarily the current directory), i.e. the
last revision in which this directory was changed.

.B %p
is not implemented (it makes no sense with Subversion).

.B %u
and
.B %m
are not implemented.

.SH CVS SUPPORT

.B vcprompt
considers the current directory a CVS working dir if file
.I CVS/Entries
exists.

.B %b
(branch) is supported by reading
.I CVS/Tag;
if that file does not exist,
.B vcprompt
displays the branch as "trunk".
.B vcprompt
looks only in the current dir, not in any subdirectories, so it will
not detect mixed-branch working dirs.

.B %r
is not supported because CVS has no global revision ID.

.B %p
is not implemented (it makes no sense with CVS).

.B %u
and
.B %m
are not supported because CVS has no easy way to get that
information.

.SH FOSSIL SUPPORT

.B vcprompt
considers the current directory a Fossil working dir if either of the
files
.I _FOSSIL_
or
.I .fslckout
exist.

All format specifiers depend on running "fossil" commands, so every
operation in a Fossil working dir is slow. All format specifiers except
.B %u
depend on a single run of "fossil status", so there's no harm in using
lots of format specifiers (however, doing so
.I is
expensive for other version control systems).

Format specifier
.B %p
is not implemented.

Format specifier
.B %u
requires running "fossil extra", so has an extra penalty compared to
the other format specifiers.

.SH CONFIGURING BASH

Use command substitution to include the output of
.B vcprompt
in your prompt. For example, add
.nf
.in +2m
if [ "$PS1" ]; then
  PS1='\\u@\\h $(vcprompt -f "[%b]")\$ '
fi
.in -2m
.fi
to
.I ~/.bashrc.
You can of course use all of bash's prompt escape sequences; the above
is just an example.

.SH CONFIGURING ZSH
Enable the
.B PROMPT_SUBST
option, and then use command substitution in
.B PROMPT
to run
.B vcprompt
every time zsh generates the prompt. For example, add
.nf
.in +2m
if [ -n "$PROMPT" ]; then
  setopt prompt_subst
  PROMPT='%n@%m $(vcprompt -f "[%b]")$ '
fi
.in -2m
.fi
to your
.I ~/.zshrc
file. You can of course use all of zsh's prompt escapes in
.B PROMPT.

.SH ENVIRONMENT
.IP VCPROMPT_FORMAT
Specifies the default format string (overridden by -f option).

.SH AUTHOR
vcprompt was written by Greg Ward <greg at gerg dot ca>.

The latest version is available from either of my public Mercurial
repositories:
.nf
.in +2m
http://hg.gerg.ca/vcprompt/
http://bitbucket.org/gward/vcprompt/
.in -2m
.fi


src/vcprompt.c
/*
 * Copyright (C) 2009-2013, Gregory P. Ward and contributors.
 *
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 */

#include "../config.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/time.h>
#include <unistd.h>
#include <signal.h>
#include <errno.h>
#include <limits.h>

#include "common.h"
#include "cvs.h"
#include "git.h"
#include "hg.h"
#include "svn.h"
#include "fossil.h"
/*
#include "bzr.h"
*/

static char* features[] = {
    /* Some version control systems don't change their working copy
       format every couple of versions (or they are just
       unmaintained), so we don't need versioned feature strings. */
    "cvs",
    "hg",
    "git",
    "fossil",

    /* Support for Subversion up to 1.6 is unconditional, because those
       versions don't require any additional libraries. Subversion >= 1.7
       requires SQLite, so it's conditional. */
    "svn-1.3",
    "svn-1.4",
    "svn-1.5",
    "svn-1.6",
#if HAVE_SQLITE3
    "svn-1.7",
    "svn-1.8",
#endif
    0,
};

#define DEFAULT_FORMAT "[%n:%b] "

void
parse_args(int argc, char** argv, options_t *options)
{
    int opt;
    while ((opt = getopt(argc, argv, "hf:dt:F")) != -1) {
        switch (opt) {
            case 'f':
                options->format = optarg;
                break;
            case 'd':
                options->debug = 1;
                break;
            case 't':
                options->timeout = strtol(optarg, NULL, 10);
                break;
            case 'F':
                options->show_features = 1;
                break;
            case 'h':
            default:
                printf("usage: %s [-h] [-d] [-t timeout_ms] [-f FORMAT]\n", argv[0]);
                printf("FORMAT (default=\"%s\") may contain:\n%s",
                DEFAULT_FORMAT,
                "  %n  show VC name\n"
                "  %b  show branch\n"
                "  %r  show revision\n"
                "  %p  show patch name (MQ, guilt, ...)\n"
                "  %u  indicate unknown (untracked) files\n"
                "  %m  indicate uncommitted changes (modified/added/removed)\n"
                "  %%  show '%'\n"
                );
                printf("Environment Variables:\n"
                "  VCPROMPT_FORMAT\n"
                );
                exit(1);
        }
    }
}

void
show_features(void)
{
    for (char **f = features; *f != NULL; f++) {
        puts(*f);
    }
}

void
parse_format(options_t *options)
{
    size_t i;

    options->show_branch = 0;
    options->show_revision = 0;
    options->show_patch = 0;
    options->show_unknown = 0;
    options->show_modified = 0;

    char *format = options->format;
    size_t len = strlen(format);
    for (i = 0; i < len; i++) {
        if (format[i] == '%') {
            i++;
            switch (format[i]) {
                case '\0':              /* at end of string: ignore */
                    break;
                case 'n':               /* name of VC system */
                    break;
                case 'b':
                    options->show_branch = 1;
                    break;
                case 'r':
                    options->show_revision = 1;
                    break;
                case 'p':
                    options->show_patch = 1;
                    break;
                case 'u':
                    options->show_unknown = 1;
                    break;
                case 'm':
                    options->show_modified = 1;
                    break;
                case '%':
                    break;
                default:
                    fprintf(stderr,
                            "error: invalid format string: %%%c\n",
                            format[i]);
                    break;
            }
        }
    }
}

void
print_result(vccontext_t *context, options_t *options, result_t *result)
{
    size_t i;
    char *format = options->format;
    size_t len = strlen(format);

    for (i = 0; i < len; i++) {
        if (format[i] == '%') {
            i++;
            switch (format[i]) {
                case '0':               /* end of string */
                    break;
                case 'n':
                    fputs(context->name, stdout);
                    break;
                case 'b':
                    if (result->branch != NULL)
                        fputs(result->branch, stdout);
                    break;
                case 'r':
                    if (result->revision != NULL)
                        fputs(result->revision, stdout);
                    break;
                case 'p':
                    if (result->patch != NULL)
                        fputs(result->patch, stdout);
                case 'u':
                    if (result->unknown)
                        putc('?', stdout);
                    break;
                case 'm':
                    if (result->modified)
                        putc('+', stdout);
                    break;
                case '%':               /* escaped % */
                    putc('%', stdout);
                    break;
                default:                /* %x printed as x */
                    putc(format[i], stdout);
            }
        }
        else {
            putc(format[i], stdout);
        }
    }
}

vccontext_t*
probe_all(vccontext_t** contexts, int num_contexts)
{
    int idx;
    for (idx = 0; idx < num_contexts; idx++) {
        vccontext_t *ctx = contexts[idx];
        if (ctx->probe(ctx)) {
            return ctx;
        }
    }
    return NULL;
}

/* walk up the directory tree until the probes work or we hit / */
vccontext_t*
probe_dirs(vccontext_t** contexts, int num_contexts)
{
    char *start_dir = malloc(PATH_MAX);
    if (getcwd(start_dir, PATH_MAX) == NULL) {
        debug("getcwd() failed: %s", strerror(errno));
        free(start_dir);
        return NULL;
    }
    char *rel_path = start_dir + strlen(start_dir);

    vccontext_t *context = NULL;
    while (1) {
        context = probe_all(contexts, num_contexts);
        if (context != NULL) {
            break;
        }
        if (rel_path == start_dir + 1) {
            debug("reached the root: %s not under version control", start_dir);
            break;
        }

        debug("no context claimed current dir: walking up the tree");
        if (-1 == chdir("..")) {
            debug("chdir(\"..\") failed: %s", strerror(errno));
            break;
        }
        do {
            rel_path--;
        } while (rel_path > start_dir && rel_path[-1] != '/');
    }
    if (context != NULL) {
        debug("found a context: %s (rel_path=%s)", context->name, rel_path);
        context->rel_path = strdup(rel_path);
    }
    free(start_dir);
    return context;
}

/* The signal handler just clears the flag and re-enables itself.  */
void
exit_on_alarm(int sig)
{
    printf("[timeout]");
    exit(1);
}

unsigned int
set_alarm(unsigned int milliseconds)
{
    struct itimerval old, new;
    new.it_interval.tv_usec = 0;
    new.it_interval.tv_sec = 0;
    new.it_value.tv_usec = 1000 * (long int) milliseconds;
    new.it_value.tv_sec =  0;
    if (setitimer(ITIMER_REAL, &new, &old) < 0)
        return 0;
    else
        return old.it_value.tv_sec;
}

int
main(int argc, char** argv)
{
    int status = 0;

    /* Establish a handler for SIGALRM signals.  */
    signal(SIGALRM, exit_on_alarm);

    char *format = getenv("VCPROMPT_FORMAT");
    if (format == NULL)
        format = DEFAULT_FORMAT;
    options_t options = {
        .debug         = 0,
        .format        = format,
        .show_branch   = 0,
        .show_revision = 0,
        .show_unknown  = 0,
        .show_modified = 0,
        .show_features = 0,
    };

    parse_args(argc, argv, &options);
    if (options.show_features) {
        show_features();
        return 0;
    }

    parse_format(&options);
    set_options(&options);

    if (options.timeout) {
        debug("will timeout after %d ms", options.timeout);
        set_alarm(options.timeout);
    } else {
        debug("will never timeout");
    }

    vccontext_t *contexts[] = {
        /* ordered by popularity, so the common case is fast */
        get_git_context(&options),
        get_hg_context(&options),
        get_svn_context(&options),
        get_cvs_context(&options),
        get_fossil_context(&options),
    };
    int num_contexts = sizeof(contexts) / sizeof(vccontext_t*);

    result_t *result = NULL;
    vccontext_t *context = NULL;

    /* Starting in the current dir, walk up the directory tree until
       someone claims that this is a working copy. */
    context = probe_dirs(contexts, num_contexts);

    /* Nobody claimed it: bail now without printing anything. */
    if (context == NULL) {
        goto done;
    }

    /* Analyze the working copy metadata and print the result. */
    result = context->get_info(context);
    if (result != NULL) {
        print_result(context, &options, result);
        free_result(result);
        if (options.debug)
            putc('\n', stdout);
    }

 done:
    for (int i = 0; i < num_contexts; i++) {
        free_context(contexts[i]);
    }
    return status;
}
